File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        map<char, char>parenthess_map{
            {'(',')'}, {'{','}'},  {'[',']'}
        };

        stack<char> char_stack;
        char tmp_char,cmp_char;
        

        for (size_t i = 0; i < s.size(); i++)
        {
            tmp_char = s[i];
            
            if(tmp_char == '('|| tmp_char == '{' || tmp_char == '[')
                char_stack.push(tmp_char);
            else{
                if(char_stack.empty())
                    return false;
                cmp_char = char_stack.top();
                char_stack.pop();
                if(parenthess_map[cmp_char] != tmp_char)
                    return false;
            }
        }
        
        if( char_stack.empty())
            return true;
        else
            return false;
         
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        def is_valid_pair(char1, char2):
            if char1 == '(':
                return char2 == ')'

            if char1 == '[':
                return char2 == ']'

            if char1 == '{':
                return char2 == '}'

        for char in s:
            if char in ['(', '[', '{']:
                stack.append(char)

            else:
                if not stack:
                    return False
                
                left_part = stack.pop()
                if not is_valid_pair(left_part, char):
                    return False
        
        return len(stack) == 0

        